connectors/src/connectors/slack/temporal/activities.ts

```diff
@@ -0,0 +1,479 @@
+  );
+  const documentId = `${channelName}-threaded-${threadTs}`;
+
+  const firstMessage = allMessages[0];
+  let sourceUrl: string | undefined = undefined;
+  const createdAt = firstMessage?.ts
+    ? parseInt(firstMessage.ts, 10) * 1000
+    : undefined;
+  if (firstMessage && firstMessage.ts) {
+    const linkRes = await client.chat.getPermalink({
+      channel: channelId,
+      message_ts: firstMessage.ts,
+    });
+    if (linkRes.ok && linkRes.permalink) {
+      sourceUrl = linkRes.permalink;
+    }
+  }
+
+  await upsertToDatasource(
+    dataSourceConfig,
+    documentId,
+    text,
+    sourceUrl,
+    createdAt
+  );
+}
+
+async function processMessageForMentions(
+  message: string,
+  connectorId: string,
+  slackClient: WebClient
+): Promise<string> {
+  const matches = message.match(/<@[A-Z-0-9]+>/g);
+  if (!matches) {
+    return message;
+  }
+  for (const m of matches) {
+    const userId = m.replace(/<|@|>/g, "");
+    const userName = await getUserName(userId, connectorId, slackClient);
+    if (!userName) {
+      continue;
+    }
+
+    message = message.replace(m, userName);
+
+    continue;
+  }
+
+  return message;
+}
+
+async function formatMessagesForUpsert(
+  channelId: string,
+  messages: Message[],
+  connectorId: string,
+  slackClient: WebClient
+) {
+  return (
+    await Promise.all(
+      messages.map(async (message) => {
+        const text = await processMessageForMentions(
+          message.text as string,
+          connectorId,
+          slackClient
+        );
+
+        const userName = await getUserName(
+          message.user as string,
+          connectorId,
+          slackClient
+        );
+        const messageDate = new Date(parseInt(message.ts as string, 10) * 1000);
+        const messageDateStr = formatDateForUpsert(messageDate);
+
+        return `>> @${userName} [${messageDateStr}]:\n ${text}\n`;
+      })
+    )
+  ).join("\n");
+}
+
+async function upsertToDatasource(
+  dataSourceConfig: DataSourceConfig,
+  documentId: string,
+  documentContent: string,
+  sourceUrl?: string,
+  createAt?: number,
+  retries = 1,
+  waitBetweenRetries = 1000
+) {
+  if (retries < 1) {
+    throw new Error("retries must be greater than 0");
+  }
+  const errors = [];
+  for (let i = 0; i < retries; i++) {
+    try {
+      const res = await _upsertToDatasource(
+        dataSourceConfig,
+        documentId,
+        documentContent,
+        sourceUrl,
+        createAt
+      );
+      return res;
+    } catch (e) {
+      errors.push(e);
+      await new Promise((resolve) => setTimeout(resolve, waitBetweenRetries));
+    }
+  }
+  throw new Error(errors.join("\n"));
+}
+
+async function _upsertToDatasource(
+  dataSourceConfig: DataSourceConfig,
+  documentId: string,
+  documentContent: string,
+  sourceUrl?: string,
+  createAt?: number
+) {
+  const dust_url = `${FRONT_API}/api/v1/w/${dataSourceConfig.workspaceId}/data_sources/${dataSourceConfig.dataSourceName}/documents/${documentId}`;
```

Comment:
need encodeURIComponent around documentId. But +1 to @fontanierh for mutualization